packages/core/src/services/chains/run.ts

```diff
@@ -103,175 +114,63 @@ async function iterate({
   configOverrides?: ConfigOverrides
 }) {
   const prevText = previousResponse?.text
-  const stepValidator = new ChainValidator({ prevText, chain, providersMap })
-  const step = await stepValidator.call().then((r) => r.unwrap())
+  const chainValidator = new ChainValidator({
+    workspace,
+    prevText,
+    chain,
+    providersMap,
+    configOverrides,
+  })
+  const streamConsumer = new ChainStreamConsumer({
+    controller,
+    previousCount,
+    documentLogUuid,
+  })
 
   try {
-    const newMessagesInStep = step.conversation.messages.slice(previousCount)
-    const sentCount = previousCount + newMessagesInStep.length
-    publishStepStartEvent(controller, step, newMessagesInStep, documentLogUuid)
-
+    const step = await chainValidator.call().then((r) => r.unwrap())
     const providerProcessor = new ProviderProcessor({
       source,
       documentLogUuid,
       config: step.config,
       apiProvider: step.provider,
       messages: step.conversation.messages,
+      saveSyncProviderLogs: step.chainCompleted,
     })
+    const { messageCount, stepStartTime } = streamConsumer.start({ step })
     const aiResult = await ai({
-      workspace,
       messages: step.conversation.messages,
       config: step.config,
       provider: step.provider,
-      schema: getSchemaForAI(step, configOverrides),
-      output: getOutputForAI(step, configOverrides),
+      schema: step.schema,
+      output: step.output,
     })
-
-    await streamAIResult(controller, aiResult)
+    const aiStreamConsumedResult = await streamConsumer.consumeAI(aiResult)
```

Comment:
not obvious what this consumeAI method is doing nor why is it not a separate service from stream consumer

Reply:
`Stream consumer` has a method `consume AI` 😅 How it would be more clear. 

Not separating it why?

Reply:
picture that this class has 4 different methods and you just called a start method (which i don't know what it does either), in this context is hard to understand why you also need a consumeAI method

Reply:
I'm not saying to change anything just expressing what I thought when looking at this for you to consider if there could be a better way of organizing the code, there might not be

Reply:
What would be a better way of organizing all of this? What I like about the class approach is that I can go to the definition to streamConsumer and understand it has 4 public methods

Reply:
I moved out this class `consumeAI`. Now this class do

1. Stream `.setup`
2. Stream `.stepCompleted`
3. Stream `.chainCompleted`